Allow more "names" as vector components

marceloandrioni opened the pull request:
Hello, this PR fix the switched eastward/northward variables in `mdal/frmts/mdal_gdal.cpp`, as discussed in #444 .
Allow more variables names to be recognized as valid vector components, e.g.:
U/X components: eastward and zonal
V/Y components:  northward and meridional

A better way would be to use regex to check valid variables names (e.g.: (x|u|eastward|zonal)) to allow for the multitude of possible long names, e.g.:

* eastward wind
* zonal wind component
* zonal component of wind
* u-wind
* x-component of wind
* ...

and [standard names](https://cfconventions.org/Data/cf-standard-names/current/build/cf-standard-name-table.html), e.g.:

* eastward_wind
* geostrophic_eastward_wind
* x_wind
* ...

but unfortunately I am not proficient in C++.

Thank you.

uclaros reviewed (commented):
Looks good, got one question though.

marceloandrioni commented on mdal/frmts/mdal_ugrid.cpp:
```diff
@@ -531,16 +531,22 @@ void MDAL::DriverUgrid::parseNetCDFVariableMetadata( int varid,
       variableName = standardName;
-      if ( MDAL::contains( standardName, "_x_" ) )
+      if ( MDAL::contains( standardName, "_x_" ) ||
+           MDAL::contains( standardName, "eastward_" ) )
       {
         *isVector = true;
         name = MDAL::replace( standardName, "_x_", "" );
+        name = MDAL::replace( name, "_eastward_", "" );
+        name = MDAL::replace( name, "eastward_", "" );
       }
-      else if ( MDAL::contains( standardName, "_y_" ) )
+      else if ( MDAL::contains( standardName, "_y_" ) ||
+                MDAL::contains( standardName, "northward_" ) )
       {
         *isVector = true;
         *isX = false;
         name = MDAL::replace( standardName, "_y_", "" );
+        name = MDAL::replace( name, "_northward_", "" );
+        name = MDAL::replace( name, "northward_", "" );
```
Hello @uclaros, in theory, all standard_names must come from the [Climate Forecast Standard Name Table](https://cfconventions.org/Data/cf-standard-names/current/build/cf-standard-name-table.html).

This table only allows x/y and eastward/northward as vector components, with the following definitions:

* "Eastward" indicates a vector component which is positive when directed eastward (negative westward).
* "Northward" indicates a vector component which is positive when directed northward (negative southward).
* "x" indicates a vector component along the grid x-axis, positive with increasing x.
* "y" indicates a vector component along the grid y-axis, positive with increasing y.

For long_names, there is no absolute rule, so the most common additions (u/x/zonal/eastaward and v/y/meridional/northward) should be used. But I am 100% sure that there is some very "creative" long names (e.g.: "wind going North") that will not be covered by this PR.

Greetings

uclaros commented on mdal/frmts/mdal_ugrid.cpp:
```diff
@@ -568,18 +574,62 @@ void MDAL::DriverUgrid::parseNetCDFVariableMetadata( int varid,
   else
   {
     variableName = longName;
-    if ( MDAL::contains( longName, ", x-component" ) || MDAL::contains( longName, "u component of " ) )
+    if ( MDAL::contains( longName, "x-component" ) ||
+         MDAL::contains( longName, "x component" ) ||
+         MDAL::contains( longName, "u-component" ) ||
+         MDAL::contains( longName, "u component" ) ||
+         MDAL::contains( longName, "eastward" ) ||
+         MDAL::contains( longName, "zonal" ) )
     {
       *isVector = true;
-      name = MDAL::replace( longName, ", x-component", "" );
-      name = MDAL::replace( name, "u component of ", "" );
+      name = MDAL::replace( longName, "x-component of", "" );
```
Skipping the leading comma causes `MeshUgridTest.DFlow12RivierGridMap` to fail

uclaros reviewed (changes requested):
mdal_gdal_netcdf_test also fails. It probably needs to be adjusted to the flipped northward/eastward handling.

marceloandrioni commented:
I saw that the tests are failing with the last commit that changes some things in the `mdal/frmts/mdal_ugrid.cpp` file.

I am not sure it's the reason, but I removed the spaces from the strings match and introduced a trim at the end to account for the extra spaces left, e.g.:
```
name = MDAL::replace( name, "u component of ", "" );
```
replaced with
```
name = MDAL::replace( name, "u component of", "" );
name = MDAL::trim( name );
```

Also, there was a check for `, x-component` and `, y-component`. I thought  the comma was a typo, so replaced with ` x-component` and `y-component`, but checking the ugrid files in the tests folder now I see that there is a lot of files using this name, e.g.:

./1dtest/dflow1d_map.nc
		mesh1d_ucx:long_name = "Flow element center velocity vector, x-component" ;
./D-Flow3/sampleNew.nc
		mesh2d_ucx:long_name = "Flow element center velocity vector, x-component" ;
./special@char/àeé.nc
		mesh1d_ucx:long_name = "Flow element center velocity vector, x-component" ;
		mesh2d_ucx:long_name = "Flow element center velocity vector, x-component" ;
./D-Flow1.1/simplebox_hex7_map.nc
		mesh2d_ucx:long_name = "Flow element center velocity vector, x-component" ;
./D-Flow1.1/manzese_1d2d_small_map.nc
		mesh1d_ucx:long_name = "Flow element center velocity vector, x-component" ;
		mesh2d_ucx:long_name = "Flow element center velocity vector, x-component" ;
./D-Flow1.2/bw_11_zonder_riviergrid_met_1dwtg_map.nc
		Mesh2D_ucx:long_name = "Flow element center velocity vector, x-component" ;

I will to fix this. Sorry for the trouble.

marceloandrioni commented:
After the last commit, now only the `mdal_gdal_netcdf_test` is failing.

```
97% tests passed, 1 tests failed out of 30

Total Test time (real) =   8.55 sec

The following tests FAILED:
	 17 - mdal_gdal_netcdf_test (Failed)
Errors while running CTest
```

The error is probably due to the switched variables eastward/northward.

```
17: Expected equality of these values:
17:   -0.4306640625
17:   value
17:     Which is: 0.0673828125
```

Does this project has any rule about only the main developers changing tests or can I jut do another push with the "new" corrected value in the [test file](https://github.com/lutraconsulting/MDAL/blob/master/tests/test_gdal_netcdf.cpp#L64)?

uclaros commented:
Thanks for taking care of that, feel free to update the test. Changing  `double value = getValueX( ds, 144 );` to `double value = getValueY( ds, 144 );` should do it :)

marceloandrioni commented:
Done, all tests are OK now.

```
100% tests passed, 0 tests failed out of 30
```

Thank you.
